### ISSUE TITLE
Why the program throw not enough memory when writting model?

### ISSUE THREAD

**Issue by maplewizard** [OpenNMT/OpenNMT#20]
I think the model has been fully built in memory already. But why it throw this error in the saving phase? How can I solve this problem?

ps. I have enough disk space.

```
th preprocess.lua -train_src mydata/src-train.txt -train_tgt mydata/tgt-train.txt -valid_src mydata/src-val.txt -valid_tgt mydata/tgt-val.txt -save_data mydata/demo200w -seq_length 100 -src_vocab_size 300000 -tgt_vocab_size 100000
Building source vocabulary...
Created dictionary of size 300004 (pruned from 616963)

Building target vocabulary...
Created dictionary of size 100004 (pruned from 208106)

Preparing training data...
... 100000 sentences prepared
... 200000 sentences prepared
... 300000 sentences prepared
... 400000 sentences prepared
... 500000 sentences prepared
... 600000 sentences prepared
... 700000 sentences prepared
... 800000 sentences prepared
... 900000 sentences prepared
... 1000000 sentences prepared
... 1100000 sentences prepared
... 1200000 sentences prepared
... 1300000 sentences prepared
... 1400000 sentences prepared
... 2200000 sentences prepared
... 2300000 sentences prepared
... 2400000 sentences prepared
... shuffling sentences
... sorting sentences by size
Prepared 2436170 sentences (8332 ignored due to length == 0 or > 100)

Preparing validation data...
... shuffling sentences
... sorting sentences by size
Prepared 9968 sentences (32 ignored due to length == 0 or > 100)

Saving source vocabulary to 'mydata/demo200w.src.dict'...
Saving target vocabulary to 'mydata/demo200w.tgt.dict'...
Saving data to 'mydata/demo200w-train.t7'...
/home/Programs/bin/luajit: not enough memory
```

**Comment by guillaumekln** [comment]
This a known limitation with LuaJIT which has a low memory limit ([1-2GB](http://stackoverflow.com/questions/35155444/why-is-luajits-memory-limited-to-1-2-gb-on-64-bit-platforms)).

The quick workaround for this issue is to use the default Lua interpreter which can be easily achieve when installing Torch:

```
TORCH_LUA_VERSION=LUA52 ./install.sh
```

We are looking for ways to mitigate this issue during the preprocessing.

**Comment by srush** [comment]
You may also be able to workaround by replacing `th` with `lua` on the command above.

**Comment by maplewizard** [comment]
@srush If I use lua instead of th, it throw another error:

```
$lua preprocess.lua -train_src vipweibo/src-train.txt -train_tgt vipweibo/tgt-train.txt -valid_src vipweibo/src-val.txt -valid_tgt vipweibo/tgt-val.txt -save_data vipweibo/demo200w -seq_length 100 -src_vocab_size 300000 -tgt_vocab_size 100000
lua: /home/sail/Programs/th/share/lua/5.1/torch/init.lua:13: cannot load '/home/sail/Programs/th/lib/lua/5.1/libtorch.so'
stack traceback:
        [C]: in function 'require'
        /home/sail/Programs/th/share/lua/5.1/torch/init.lua:13: in main chunk
        [C]: in function 'require'
        /home/sail/Programs/th/share/lua/5.1/nn/init.lua:1: in main chunk
        [C]: in function 'require'
        /home/sail/Programs/th/share/lua/5.1/nngraph/init.lua:1: in main chunk
        [C]: in function 'require'
        ./onmt/modules/Sequencer.lua:1: in main chunk
        [C]: in function 'require'
        ./onmt/modules/init.lua:3: in main chunk
        [C]: in function 'require'
        ./onmt/init.lua:3: in main chunk
        [C]: in function 'require'
        preprocess.lua:1: in main chunk
        [C]: ?
```

**Comment by srush** [comment]
(BTW I would really like to find a workaround here. This is an annoying issue.)

**Comment by guillaumekln** [comment]
There should be some improvements with the above commit. `torch.save` is now told to not keep track of serialized objects as it has a memory cost. See this [documentation](https://github.com/torch/torch7/blob/master/doc/file.md#referencedref) for reference.

**Comment by jsenellart-systran** [comment]
@maplewizard : is ac658383567243f7669b3d6d1525c0a761bb1285 fixing the problem for you?

**Comment by guillaumekln** [comment]
Fixed by #29.
